### PR TITLE
[UPSTREAM] TGUI Say

### PR DESCRIPTION
**Pull request by Bobbanz1**
## About The Pull Request
This PR ports TGUI Say from Beestation

It accomplishes this by pulling from upstream the following PR: 
* https://github.com/BeeStation/BeeStation-Hornet/pull/8404

That PR in turn ports TGUI Say from TG Station, those are in particular the following PRs:
* https://github.com/tgstation/tgstation/pull/67116
* https://github.com/tgstation/tgstation/pull/67971
* https://github.com/tgstation/tgstation/pull/67137
* https://github.com/tgstation/tgstation/pull/67265

What the Beestation port does is it implements TGUI Say. However it does it in such a way that allows the player to choose between either using the TGUI Say interface or the original interface. 

The PR also introduces the ability to let the player choose between using TGUI Inputs, such as tgui buttons/tgui lists and the original interface.

Additional Changes ported from Beestations TGUI Say PR:
* Added explo (although we don't use it), uplink, `:l`, `:r` and `:h` channel handling
* Added support for `.channel` prefixes
* Colors were tweaked
* The entire new window is draggable, excluding the textbox.
* The border has been tweaked slightly, it is not animated but has a minor gradient on both the top and bottom
* Added a border around the text input part of the box.
* Added LOOC channel (Quickly accessible by pressing L on your keyboard)
* Legacy support pref, with typing indicators and all
* Typing `;` in Say mode will convert the radio prefix to `;`, and removing it will go back to Say.
* Shift+TAB and right clicking the button will go back in channels.
* Re-opening the window on a different channel will no longer break the button contents.
* Added "show prefix" mode which treats `:s` as text in the input box while still updating the color and button.
* Pressing ESC on the game window with the say unfocused will exit the say.

SQL Update that needs to be done before test-merge and/or full merge
```sql
UPDATE SS13_preferences SET preference_value = preference_value | (1<<14) | (1<<15) | (1<<16) | (1<<17) WHERE preference_tag = 2;
```

Todo:
* [ ] Figure out how to record this stuff in action without reaching the github's file size limit

## Why It's Good For The Game
Keeps the codebase somewhat up to date with Beestation. 
This also makes it easier to import a future pull request from Beestation if we have the preferences and code required for it already in the codebase.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

**TGUI Say Enabled:**
TGUISayPreferences
withTGUISayEnabled

**TGUI Say Disabled:**
TGUISayDisabled
withoutTGUISayEnabled

**TGUI Input:**
_Enabled with Large Buttons_
tguiinputsettingslargeokcancel
TGUIInputListExampleTwo

_Enabled with Small Buttons and Button Locations Swapped_
TGUIInputSettingsSmallButtonCancelOK
TguiListStuff

_Disabled_
TguiInputSettingsDisabled
TguiInputDisabled

**How the Preference section looks for TGUI**
Skärmbild 2024-03-14 091809

**How the various channels look alongside their associated keys/prefixes:**
_Say (T Key on Keyboard)_
withTGUISayEnabled

_Me (M Key on Keyboard)_
me

_LOOC (L Key on Keyboard)_
l_looc

_OOC (O Key on Keyboard)_
o_ooc

_Common_
common

_Engineering (Prefix: :e or .e)_
engineering

_Security_
security

_Supply_
supply

_Command_
command

_Service (Prefix: :v or .v)_
service

_Science_
science

_Hologram_
holo

_Medical_
medical

_Munitions_
munitions

_ATC_
atc

_Centcom_
centcom

_Syndicate_
syndicate

_Pirates_
space_pirate

</details>

## Changelog
:cl:itsmeow
add: Added TGUI-Say (can be disabled in prefs, light theme also in prefs) - Tab to switch channels.
add: Added new TGUI Say keybinds - Y for Radio, L for LOOC (not currently rebindable).
add: Added TGUI input preferences (mostly unused currently).
add: Added glorfing to TGUI-Say - if you get hit hard enough, your message will be cut off with "-OW!!!", "-AUGH!!!", "-GLORF!!" or similar.
admin: Added logging to typing indicators.
add: Added OOC Verb "Refresh TGUI".
tweak: Legacy say hotkey macros (Ctrl+T, Ctrl+O) have been replaced with T and O.
/:cl:

### CODE REVIEW

**Comment by Bokkiewokkie** [comment]
Want to add configurable hotkeys for chat messages; z being a permanent hotkey means qwertz keyboard users are getting their keybinds mixed up.

File: tgui/docs/tgui-for-custom-html-popups.md [hunk 1]
```diff
@@ -257,3 +257,280 @@ Byond.winget(null, 'url').then((serverUrl) => {
   Byond.command('.quit');
 });
 ```
+# Using TGUI and Byond API for custom HTML popups
```
Comment (DeltaFire15) [inline tgui/docs/tgui-for-custom-html-popups.md:260]: This is doubling what this file already says.

**Comment by Bobbanz1** [comment]
> Want to add configurable hotkeys for chat messages; z being a permanent hotkey means qwertz keyboard users are getting their keybinds mixed up.

I can see what I can do when I get the time

File: tgui/docs/tgui-for-custom-html-popups.md (hunk 1 above)
Comment (Bobbanz1) [reply tgui/docs/tgui-for-custom-html-popups.md]: Fixed

**Review by Bokkiewokkie — APPROVED** [review]
We can do the hotkey update in a next PR, so we don't have this open forever

**Pull request merged**